File: app/src/main/java/com/example/bridgeapp/structure/PlayObject.java
package com.example.bridgeapp.structure;

import android.os.Parcel;
import android.os.Parcelable;

import com.example.bridgeapp.logic.BidStageState;
import com.example.bridgeapp.logic.DealStageState;
import com.example.bridgeapp.logic.PointStageState;
import com.example.bridgeapp.util.GameStage;

import java.util.List;

public class PlayObject implements Parcelable {

    private final int dealingPlayer;

    private GameStage stage;
    private DealStageState dealHistory;
    private BidStageState bidHistory;
    private PointStageState pointHistory;
    private boolean noPointPlay;

    public PlayObject(int dealingPlayer) {
        this.dealingPlayer = dealingPlayer;
        stage = GameStage.DEAL;
        noPointPlay = false;
    }

    public static final Creator<PlayObject> CREATOR = new Creator<PlayObject>() {
        @Override
        public PlayObject createFromParcel(Parcel in) {
            return new PlayObject(in);
        }

        @Override
        public PlayObject[] newArray(int size) {
            return new PlayObject[size];
        }
    };

    public void setStage(GameStage stage) {
        this.stage = stage;
    }

    public GameStage getStage() {
        return stage;
    }

    public void setDealHistory(DealStageState dealHistory) {
        this.dealHistory = dealHistory;
    }

    public void setBidHistory(BidStageState bidHistory){
        this.bidHistory = bidHistory;
    }

    public void setPointHistory(PointStageState pointHistory){
        this.pointHistory = pointHistory;
    }

    public DealStageState getDealHistory() {
        return dealHistory;
    }

    public BidStageState getBidHistory() {
        return bidHistory;
    }

    public PointStageState getPointHistory() {
        return pointHistory;
    }

    public HandObject[] getHands() {
        return this.dealHistory.getHands();
    }

    public List<BidObject> getBids(){
        return this.bidHistory.getBidList();
    }

    public ContractObject getContract() {
        return this.bidHistory.getContract();
    }

    public int getDealingPlayer() {
        return dealingPlayer;
    }

    public void setNoPoint(boolean noPoint) {
        this.noPointPlay = noPoint;
    }

    public boolean isNoPointPlay() {
        return noPointPlay;
    }

    public PlayObject(Parcel in) {
        this.dealingPlayer = in.readInt();
        this.stage = GameStage.valueOf(in.readString());
        this.dealHistory = in.readTypedObject(DealStageState.CREATOR);
        this.bidHistory = in.readTypedObject(BidStageState.CREATOR);
        this.pointHistory = in.readTypedObject(PointStageState.CREATOR);
        this.noPointPlay = in.readByte() != 0;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(this.dealingPlayer);
        dest.writeString(this.stage.name());
        dest.writeTypedObject(this.dealHistory, 0);
        dest.writeTypedObject(this.bidHistory, 0);
        dest.writeTypedObject(this.pointHistory, 0);
        dest.writeByte((byte) (noPointPlay ? 1 : 0));
    }
}

File: app/src/main/java/com/example/bridgeapp/structure/PlayObject.kt
package com.example.bridgeapp.structure

import android.os.Parcel
import android.os.Parcelable
import com.example.bridgeapp.logic.BidStageState
import com.example.bridgeapp.logic.DealStageState
import com.example.bridgeapp.logic.PointStageState
import com.example.bridgeapp.util.GameStage

class PlayObject : Parcelable {
    val dealingPlayer: Int
    var stage: GameStage
    var dealHistory: DealStageState? = null
    var bidHistory: BidStageState? = null
    var pointHistory: PointStageState? = null
    var isNoPointPlay: Boolean
        private set

    constructor(dealingPlayer: Int) {
        this.dealingPlayer = dealingPlayer
        stage = GameStage.DEAL
        isNoPointPlay = false
    }

    val hands: Array<HandObject?>?
        get() = dealHistory!!.hands

    val bids: List<BidObject?>?
        get() = bidHistory!!.bidList

    val contract: ContractObject?
        get() = bidHistory!!.contract

    fun setNoPoint(noPoint: Boolean) {
        isNoPointPlay = noPoint
    }

    constructor(`in`: Parcel) {
        dealingPlayer = `in`.readInt()
        stage = GameStage.valueOf(`in`.readString()!!)
        dealHistory = `in`.readTypedObject(DealStageState.CREATOR)
        bidHistory = `in`.readTypedObject(BidStageState.CREATOR)
        pointHistory = `in`.readTypedObject(PointStageState.CREATOR)
        isNoPointPlay = `in`.readByte().toInt() != 0
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(dealingPlayer)
        dest.writeString(stage.name)
        dest.writeTypedObject(dealHistory, 0)
        dest.writeTypedObject(bidHistory, 0)
        dest.writeTypedObject(pointHistory, 0)
        dest.writeByte((if (isNoPointPlay) 1 else 0).toByte())
    }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<PlayObject?> = object : Parcelable.Creator<PlayObject?> {
            override fun createFromParcel(`in`: Parcel): PlayObject? {
                return PlayObject(`in`)
            }

            override fun newArray(size: Int): Array<PlayObject?> {
                return arrayOfNulls(size)
            }
        }
    }
}